lib/vfscore/automount.c

```diff
@@ -0,0 +1,309 @@
+	UK_ASSERT(vv->drv);
+
+	if (!pos)
+		return;
+
+	/* Allow empty flags: "<src_dev>:<mntpoint>:<fsdriver>::<opts>" */
+	flags = vfscore_fstab_next_volume_arg(&pos);
+	if (flags && flags[0] != '\0')
+		vv->flags = strtol(flags, NULL, 0);
+	else
+		vv->flags = 0;
+
+	vv->opts = vfscore_fstab_next_volume_arg(&pos);
+	if (vv->opts && vv->opts[0] == '\0')
+		vv->opts = NULL;
+}
+#endif /* CONFIG_LIBVFSCORE_FSTAB */
+
+#ifdef CONFIG_LIBUKCPIO
+static int vfscore_mount_initrd_volume(struct vfscore_volume *vv)
+{
+	struct ukplat_memregion_desc *initrd;
+	int rc;
+
+	UK_ASSERT(vv);
+
+	/* TODO: Support multiple Initial RAM Disks */
+	rc = ukplat_memregion_find_initrd0(&initrd);
+	if (unlikely(rc < 0)) {
+		uk_pr_crit("Could not find an initrd!\n");
+
+		return -1;
+	}
+
+	uk_pr_info("Mount ramfs to /...\n");
+
+	rc = mount(vv->sdev, vv->path, "ramfs", vv->flags, vv->opts);
+	if (unlikely(rc)) {
+		uk_pr_crit("Failed to mount ramfs to /: %d\n", errno);
+
+		return -1;
+	}
+
+	uk_pr_info("Extracting initrd @ %p (%"__PRIsz" bytes) to %s...\n",
+		   (void *)initrd->vbase, initrd->len, vv->path);
+
+	rc = ukcpio_extract(vv->path, (void *)initrd->vbase, initrd->len);
+	if (unlikely(rc)) {
+		uk_pr_crit("Failed to extract cpio archive to %s: %d\n",
+			   vv->path, rc);
+
+		return -1;
+	}
+
+	return 0;
+}
+
+#else /* CONFIG_LIBUKCPIO */
+static int vfscore_mount_initrd_volume(struct vfscore_volume *vv __unused)
+{
+	uk_pr_warn("Cannot extract InitRD. Enable CONFIG_LIBRAMFS.\n");
+
+	return -1;
+}
+#endif /* !CONFIG_LIBUKCPIO */
+
+#ifdef CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS
+static int vfscore_automount_rootfs(void)
+{
+	/* Convert to `struct vfscore_volume` */
+	struct vfscore_volume vv = {
+		.sdev = rootdev,
+		.path = "/",
+		.drv = rootfs,
+		.flags = rootflags,
+		.opts = rootopts,
+	};
+	int rc;
+
+	/*
+	 * Initialization of the root filesystem '/'
+	 * NOTE: Any additional sub mount points (like '/dev' with devfs)
+	 * have to be mounted later.
+	 *
+	 * Silently return 0, as user might not have configured implicit rootfs.
+	 */
+	if (!vv.drv || vv.drv[0] == '\0')
+		return 0;
+
+#if CONFIG_LIBVFSCORE_ROOTFS_INITRD
+	if (!strncmp(vv.drv, "initrd", sizeof("initrd") - 1))
+		return vfscore_mount_initrd_volume(&vv);
+#endif /* CONFIG_LIBVFSCORE_ROOTFS_INITRD */
+
+	uk_pr_info("Mount %s to /...\n", vv.drv);
+	rc = vfscore_mount_volume(&vv);
+	if (unlikely(rc))
+		uk_pr_crit("Failed to mount %s (%s) at /: %d\n", vv.sdev,
+			   vv.drv, rc);
+
+	return rc;
+}
+#else /* CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
+static int vfscore_automount_rootfs(void)
+{
+	return 0;
+}
+#endif /* !CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS */
+
+#ifdef CONFIG_LIBVFSCORE_FSTAB
+static int vfscore_automount_fstab_volumes(void)
+{
+	struct vfscore_volume vv;
+	int rc, i;
+
+	for (i = 0; i < CONFIG_LIBVFSCORE_FSTAB_SIZE && vfscore_fstab[i]; i++) {
+		vfscore_fstab_fetch_volume_args(vfscore_fstab[i], &vv);
+
+		uk_pr_info("Mount %s to %s...\n", vv.sdev, vv.path);
```

Comment:
What is `vv.sdev` saying in the initrd case? Maybe you could place this message under the else case and print a message for initrd like: "Extracting initrd to %s...\n"

Reply:
`vv.sdev` is a mandatory argument as per `fstab` syntax:

```
vfs.fstab=[
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		...
	]
```
`vv.sdev` is therefore `<src_dev>`.

Reply:
My question was that the initrd mounting code does not need a source device as parameter. According to the code we accept anything that is in there which is actually fine. :+1: